### ISSUE TITLE
Load decrypted game updates to avoid patching game backups

### ISSUE THREAD

**Issue by LunaMoo** [hrydgard/ppsspp#12344]
Currently we have no means of updating games without patching the iso and this might be:
- confusing especially for issues like #9820, especially since many games will still load and display update "notes" which lie about being updated by showing version number from the update note even if they fail to run updated executable and people assume they're on latest version then think the DLC's just doesn't work,
- it might also cause problem when trying to play adhoc as most multiplayer centered games had meaningful content updates and might not be compatible between patched and unpatched game with small portion of the community understanding how to patch games,
- causing problems like in #11106 where updated game has an option to run "updated" for special DLC or revert to non-updated state,
- I guess it's also bad for "game preservation" as you can't keep more than 1 backup in most countries, so requiring to modify that backup is bad, it's also additional hassle in addition to having to decrypt the update using real hardware.

 Outside of potentially improving solving the problems above, it could help or be useful in some way also to that Killzone DLC ~ #12343.

 Even if we would have such system for loading decrypted updates, not many people even know we can't decrypt updates outside of real PSP, so it could also print some error on the screen when failing to decrypt update and no decrypted version of it was found.

**Comment by SamRohod** [comment]
> Even if we would have such system for loading decrypted updates, not many people even know we can't decrypt updates outside of real PSP, so it could also print some error on the screen when failing to decrypt update and no decrypted version of it was found.

Can't [PSN PKG Decryptor & Extractor](https://sites.google.com/site/theleecherman/psnpkgdecryptor-extractor) decrypt .pkg update files without real hardware?

**Comment by LunaMoo** [comment]
It can't. Decrypting update eboot is not possible without real hardware due to key for it not being public.

**Comment by MojoJojoDojo** [comment]
Could JPCSP decrypt them? seeing as it got emulates a lot of low level stuff that PPSSPP does not even comes close to.

**Comment by MojoJojoDojo** [comment]
BTW to add to all of this , the infrastructure build AN2RME made here https://github.com/hrydgard/ppsspp/issues/14256 managed to download an update for SOCOM FireTeam Bravo 2 but I guess that without the decryption keys it couldn't proceed after downloading it.
I think i still have it somewhere.

**Comment by Linblow** [comment]
See my [post here](https://github.com/hrydgard/ppsspp/issues/12343#issuecomment-1260417049) related to this issue.

**Comment by 1jtp8sobiu** [comment]
http://cmd0725.blog.fc2.com/blog-entry-3065.html

https://egg.5ch.net/test/read.cgi/software/1567678253/457

`static const u8 key_2E5E10F0_xor[] = {0x69, 0xBA, 0x55, 0x34, 0xF0, 0xC0, 0xD6, 0x71, 0xE3, 0x1F, 0xDB, 0x97, 0xE0, 0x7C, 0xD2, 0x2A};
`

**Comment by hrydgard** [comment]
@1jtp8sobiu Can't open the 2nd link.

The first one doesn't offer much clue about what that key is. It matches the name of one of the keys in PrxDecrypter, but with an added xor suffix and the bytes are not the same.

**Comment by LunaMoo** [comment]
I asked tpunix about this like 10 years ago, his reply:

> look the source of pspcipher:
{ 0x2e5e10f0, key_2e5e10f0, xor_2e5e10f0, 0x48, 5 }
for 2e5e10f0, the code is 0x48, which is the key index used by kirk cmd 7.

>we miss this key now Sad

>all support code can be found at kirk_engine.c

And sure, our kirk_engine.c does miss cmd 7 code

````
void kirk7(u8* outbuff, const u8* inbuff, size_t size, int keyId)
{
  AES_ctx aesKey;
  u8* key = kirk_4_7_get_key(keyId);
  AES_set_key(&aesKey, key, 128);
  AES_cbc_decrypt(&aesKey, inbuff, outbuff, (int)size);
}
````

but looking online our kirk_engine.c seems outdated compared to for example:
https://github.com/cielavenir/psp-savedata-endecrypter/blob/master/libkirk/kirk_engine.c#L334

so maybe all the info is out there already and the missing key is known, just needs someone to understand how to use it, I probably know less than 10 years ago about those things ^_^;

**Comment by SamRohod** [comment]
Hi @hrydgard & @LunaMoo 

Based on what you know, can you provide some information on what type of key/keys we are missing for this functionality to work. Is it actually a Kirk key? its location? or any other descriptions that might help if someone is trying to take stab at this.

**Comment by LunaMoo** [comment]
Seems like we already have the required keys, it probably needs new method, I messed around it in https://github.com/hrydgard/ppsspp/pull/18562, but unfortunately while that resulted in correct file size which might be considered slight improvement, it was still gibberish.

**Comment by Linblow** [comment]
Good news KIRK rom has been dumped thanks to a bunch of boys. It's not released yet, but it will probably be released along with the IDA plugin to RE it, or an already RE version will be released. Either way, it's just a matter of time before KIRK 2/3 exact algorithm and static keys are known. They already figured out KIRK 5/8 and 6/9.

**Comment by SamRohod** [comment]
Do these links have what's needed?
[(Link 1)](https://wololo.net/2024/01/16/the-playstation-portables-kirk-rom-has-been-dumped/)
[(Link 2)](https://www.psdevwiki.com/psp/Kirk)
[(Link 3)](https://darthsternie.net/psp-assorted-firmwares/)

**Comment by Linblow** [comment]
@SamRohod WIP

**Comment by Proletar666** [comment]
@Linblow is it still in the works?

**Comment by Linblow** [comment]
Yes, it is. Like I said before, I intend to wrap up a lot of things in 2026
